### PR TITLE
enhancement(aws_ecs_metrics source): comply with component spec

### LINKED ISSUE #10056: Instrument the `aws_ecs_metrics` source with event processing metrics
Per #9326

### PR DESCRIPTION
**Pull request by jdrouet**
update the event processing metrics

closes #10056

Signed-off-by: Jérémie Drouet <jeremie.drouet@datadoghq.com>

### CODE REVIEW

**Review by leebenson — DISMISSED** [review]
Appears to be missing a `BytesReceived` event per https://github.com/vectordotdev/vector/blob/master/docs/specs/component.md#bytesreceived

File: src/internal_events/aws_ecs_metrics.rs [hunk 1]
```diff
@@ -13,13 +13,17 @@ pub struct AwsEcsMetricsReceived {
 
 impl InternalEvent for AwsEcsMetricsReceived {
```
Comment (leebenson) [inline src/internal_events/aws_ecs_metrics.rs:14]: Per [the Events spec](https://github.com/vectordotdev/vector/blob/master/docs/specs/component.md#events), internal events indicating events received should be named in the format of `<Component>Event(s?)Received`.

In this case, `AwsEcsMetricsEventsReceived` seems accurate.

Comment (leebenson) [inline src/internal_events/aws_ecs_metrics.rs:14]: When a URL is relevant to a component, an `endpoint` should be included in the event per https://github.com/vectordotdev/vector/blob/master/docs/specs/component.md#options.

I can see this is already the case with `AwsEcsMetricsParseError`, but I think other events should include it too.

**Comment by leebenson** [comment]
A reminder to update the telemetry section in cue docs too, i.e.:

https://github.com/vectordotdev/vector/blob/15ceca96f5126193353c984df92cc08c8724228c/website/cue/reference/components/sources/aws_ecs_metrics.cue#L233-L243:

`component_received_event_bytes_total` is missing, for example.

File: src/internal_events/aws_ecs_metrics.rs [hunk 2]
```diff
@@ -6,20 +6,50 @@ use metrics::{counter, histogram};
 use vector_core::internal_event::InternalEvent;
```
Comment (bruceg) [inline src/internal_events/aws_ecs_metrics.rs:6]: Remember to drop the `skip check-events` comment at the top of the file when doing these updates.

**Review by bruceg — CHANGES_REQUESTED** [review]
Looks OK to me except for a couple clippy issues and one required change below.

File: src/internal_events/aws_ecs_metrics.rs [hunk 3]
```diff
@@ -43,53 +71,145 @@ impl InternalEvent for AwsEcsMetricsRequestCompleted {
...
+            "stage" => "sending",
+        );
+        counter!(
+            "component_discarded_events_total", self.count as u64,
+            "error" => self.error.clone(),
+            "error_type" => "stream",
+            "stage" => "sending",
+        );
     }
 }
```
Comment (bruceg) [inline src/internal_events/aws_ecs_metrics.rs:215]: Ref my comments in other PRs about these events.

**Review by bruceg — CHANGES_REQUESTED** [review]

File: src/internal_events/aws_ecs_metrics.rs [hunk 4]
```diff
@@ -43,53 +23,91 @@ impl InternalEvent for AwsEcsMetricsRequestCompleted {
...
-            url = %self.url,
+            endpoint = %self.endpoint,
             internal_log_rate_secs = 10
         );
     }
 
     fn emit_metrics(&self) {
         counter!("parse_errors_total", 1);
+        counter!(
+            "component_errors_total", 1,
```
Comment (001wwang) [inline src/internal_events/aws_ecs_metrics.rs:49]: Should this include an `error` tag? I've left this comment elsewhere so to explain my perspective -- my reading of the [component spec](https://github.com/vectordotdev/vector/blob/master/docs/specs/component.md#error) makes me feel like it should but curious if it's more discretionary. cc: @bruceg

File: src/internal_events/aws_ecs_metrics.rs [hunk 5]
```diff
@@ -43,53 +23,91 @@ impl InternalEvent for AwsEcsMetricsRequestCompleted {
...
 }
 
-impl InternalEvent for AwsEcsMetricsErrorResponse<'_> {
+impl InternalEvent for AwsEcsMetricsResponseError<'_> {
     fn emit_logs(&self) {
-        error!(message = "HTTP error response.", url = %self.url, code = %self.code);
+        error!(
+            message = "HTTP error response.",
+            endpoint = %self.endpoint,
+            code = %self.code,
```
Comment (001wwang) [inline src/internal_events/aws_ecs_metrics.rs:68]: Should the log and metric label's for `code` be renamed `error` for consistency?

File: src/internal_events/common.rs [hunk 6]
```diff
@@ -90,3 +90,36 @@ impl InternalEvent for AwsBytesSent {
...
     }
 }
+
+#[derive(Debug)]
+pub struct StreamClosedError {
+    pub error: String,
+    pub count: usize,
+}
+
+impl InternalEvent for StreamClosedError {
```
Comment (001wwang) [inline src/internal_events/common.rs:100]: Just to note: I left some comments/questions about `StreamClosedError` in the[ EventsStoreDB Metrics PR](https://github.com/vectordotdev/vector/pull/10890)

**Review by 001wwang — COMMENTED** [review]
Looks good to me overall! Included some questions that I'll wait on before approving.

File: src/internal_events/aws_ecs_metrics.rs (hunk 4 above)
Comment (jdrouet) [reply src/internal_events/aws_ecs_metrics.rs:49]: I aggree, the `error` field should be here

File: src/internal_events/aws_ecs_metrics.rs [hunk 7]
```diff
@@ -43,53 +23,92 @@ impl InternalEvent for AwsEcsMetricsRequestCompleted {
...
```
Comment (001wwang) [inline src/internal_events/aws_ecs_metrics.rs:69]: Think the update from `code` -> `error` may have been missed here.

**Review by 001wwang — APPROVED** [review]
Great! Just one minor thing that might have missed updating.

File: src/internal_events/aws_ecs_metrics.rs (hunk 7 above)
Comment (jdrouet) [reply src/internal_events/aws_ecs_metrics.rs:69]: 👍

File: src/sources/aws_ecs_metrics/mod.rs [hunk 8]
```diff
@@ -142,43 +143,57 @@ async fn aws_ecs_metrics(
                             end: Instant::now()
                         });
 
-                        let byte_size = body.len();
```
Comment (bruceg) [inline src/sources/aws_ecs_metrics/mod.rs:145]: Tiny nit: You could leave this hear, as you do the same thing inside the `match` below.

**Review by bruceg — APPROVED** [review]
LGTM. One really tiny suggestion, but it's not blocking.

File: src/sources/aws_ecs_metrics/mod.rs (hunk 8 above)
Comment (jdrouet) [reply src/sources/aws_ecs_metrics/mod.rs:145]: fixed 😉

File: src/internal_events/aws_ecs_metrics.rs [hunk 9]
```diff
@@ -1,28 +1,8 @@
...
-
-impl InternalEvent for AwsEcsMetricsReceived {
-    fn emit_logs(&self) {
-        debug!(message = "Scraped events.", ?self.count);
-    }
-
-    fn emit_metrics(&self) {
-        counter!("component_received_events_total", self.count as u64);
-        counter!("events_in_total", self.count as u64);
-        counter!("processed_bytes_total", self.byte_size as u64);
```
[the remark this answers is not in the document]
Comment (jdrouet) [reply src/internal_events/aws_ecs_metrics.rs:22]: I'll do that

[the remark this answers is not in the document]
Comment (jszwedko) [reply src/internal_events/aws_ecs_metrics.rs:22]: Thank you! ❤️

File: src/sources/aws_ecs_metrics/mod.rs [hunk 10]
```diff
@@ -144,41 +145,56 @@ async fn aws_ecs_metrics(
...
+                            protocol: "http",
+                            http_path: uri.path(),
+                        });
+
                         match parser::parse(body.as_ref(), namespace.clone()) {
                             Ok(metrics) => {
-                                emit!(&AwsEcsMetricsReceived {
+                                let count = metrics.len();
+                                emit!(&HttpEventsReceived {
                                     byte_size,
```
Comment (jszwedko) [inline src/sources/aws_ecs_metrics/mod.rs:158]: I think this `byte_size` is actually wrong too. It should be using `ByteSizeOf::size_of` to calculate the in-memory byte count.

**Pull request merged**